### PR TITLE
LibWeb: Partially implement `Element::scroll_width` and `Element::scroll_height`

### PR DESCRIPTION
**Pull request by Tyrubias**
Currently, `Element::scroll_width` and `Element::scroll_height` are unimplemented and always return 0. This pull request implements most of these functions according to [the specification](https://w3c.github.io/csswg-drafts/cssom-view/#dom-element-scrollwidth).

If possible, I would like some help finding the scrolling area of an element.

### CODE REVIEW

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 1]
```diff
@@ -1001,15 +1001,75 @@ void Element::set_scroll_top(double y)
...
+    // 1. Let document be the element's node document.
+    auto const& document = this->document();
+
+    // 2. If document is not the active document, return zero and terminate these steps.
+    if (!document.is_active())
+        return 0;
+
+    // 3. Let viewport width be the width of the viewport excluding the width of the scroll bar,
+    //    if any, or zero if there is no viewport.
+    // FIXME: Check if there is a viewport
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1016]: If `document.browsing_context()` returns null, there is no viewport. So we can use that here.

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 2]
```diff
@@ -1001,15 +1001,75 @@ void Element::set_scroll_top(double y)
...
+    if (is<HTML::HTMLBodyElement>(*this) && document.in_quirks_mode() && !this->is_potentially_scrollable())
+        return max(document.browsing_context()->viewport_scroll_offset().x(), viewport_width);
+
+    // 6. If the element does not have any associated box return zero and terminate these steps.
+    if (!this->layout_node() || !is<Layout::BlockContainer>(this->layout_node()))
+        return 0;
+
+    // FIXME: 7. Return the width of the element's scrolling area.
+    //       This may have to go in a separate PR, as it involves
+    //       CSS overflow implementation
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1036]: This meta-commentary about your PR and doesn't need to be in the code :)

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 1 above)
Comment (Tyrubias) [reply Userland/Libraries/LibWeb/DOM/Element.cpp]: Thank you! I added this check. As a separate question, how do we calculate the viewport scrolling area width if there is no viewport? The spec is not very clear on this point.

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 2 above)
Comment (Tyrubias) [reply Userland/Libraries/LibWeb/DOM/Element.cpp]: Oops,  thanks for the heads up. I changed it to a note that just warns about the need for a CSS overflow implementation before that logic can be implemented.

**Review by linusg — CHANGES_REQUESTED** [review]
Please squash the commits as per https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy

> Squash your commits when making revisions after a patch review.

**Review by linusg — COMMENTED** [review]
Please also remove the second commit message from the commit description, it's no longer relevant :)

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 3]
```diff
@@ -1001,15 +1001,73 @@ void Element::set_scroll_top(double y)
...
+    // 4. If the element is the root element and document is not in quirks mode
+    //    return max(viewport scrolling area width, viewport width).
+    if (is<HTML::HTMLHtmlElement>(*this) && !document.in_quirks_mode())
+        return max(browsing_context->viewport_scroll_offset().x(), viewport_width);
+
+    // 5. If the element is the body element, document is in quirks mode
+    //    and the element is not potentially scrollable,
+    //    return max(viewport scrolling area width, viewport width).
+    if (is<HTML::HTMLBodyElement>(*this) && document.in_quirks_mode() && !this->is_potentially_scrollable())
+        return max(browsing_context->viewport_scroll_offset().x(), viewport_width);
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1028]: Hmm, is `viewport scrolling area width` in the spec really the same as the viewport scroll offset X position? That sounds strange.

File: Userland/Libraries/LibWeb/DOM/Element.cpp [hunk 4]
```diff
@@ -1001,15 +1001,73 @@ void Element::set_scroll_top(double y)
...
+    // 4. If the element is the root element and document is not in quirks mode
+    //    return max(viewport scrolling area height, viewport height).
+    if (is<HTML::HTMLHtmlElement>(*this) && !document.in_quirks_mode())
+        return max(browsing_context->viewport_scroll_offset().y(), viewport_height);
+
+    // 5. If the element is the body element, document is in quirks mode
+    //    and the element is not potentially scrollable,
+    //    return max(viewport scrolling area height, viewport height).
+    if (is<HTML::HTMLBodyElement>(*this) && document.in_quirks_mode() && !this->is_potentially_scrollable())
+        return max(browsing_context->viewport_scroll_offset().y(), viewport_height);
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/DOM/Element.cpp:1063]: Same question but for scrolling area height.

File: Userland/Libraries/LibWeb/DOM/Element.cpp (hunk 3 above)
Comment (Tyrubias) [reply Userland/Libraries/LibWeb/DOM/Element.cpp:1028]: Oh yeah I think I made a mistake here. I'm confused — does viewport scrolling area calculation also depend on CSS overflow? I see that we do calculate the overflow for a block formatting context (which is what the root element should be), but I'm not sure I understand where we're saving that information.

**Comment by GMTA** [comment]
Build is failing

**Pull request closed without merging**